src/Algebra/Ordered/Structures.agda

```diff
@@ -0,0 +1,308 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Ordered algebraic structures (not packed up with sets, orders,
+-- operations, etc.)
+------------------------------------------------------------------------
+
+-- The contents of this module should be accessed via
+-- `Algebra.Ordered`.
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Relation.Binary.Core using (Rel; _⇒_)
+
+module Algebra.Ordered.Structures
+  {a ℓ₁ ℓ₂} {A : Set a}  -- The underlying set
+  (_≈_ : Rel A ℓ₁)       -- The underlying equality relation
+  (_≤_ : Rel A ℓ₂)       -- The order
+  where
+
+open import Algebra.Core
+open import Algebra.Definitions _≈_
+open import Algebra.Structures _≈_
+open import Algebra.Ordered.Definitions _≤_ using (Monotonic₂)
+open import Data.Product using (proj₁; proj₂)
+open import Level using (_⊔_)
+open import Relation.Binary.Definitions using (Transitive)
+open import Relation.Binary.Structures using (IsPreorder; IsPartialOrder)
+open import Relation.Binary.Consequences using (mono₂⇒cong₂)
+
+------------------------------------------------------------------------
+-- Preordered structures
+
+-- Preordered magmas (promagmas)
+
+record IsPromagma (∙ : Op₂ A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+  field
+    isPreorder  : IsPreorder _≈_ _≤_
+    ∙-cong      : Congruent₂ ∙
+    mono        : Monotonic₂ ∙
```

Comment:
It'd be nice to have `monoˡ` and `monoʳ`, too, like for `cong`.

Reply:
Good point, will add those! 👍

Reply:
Uh, now I'm confused. The type of `congˡ` is `LeftCongruent` which actually states that the operation is congruent in its second, i.e. **right** argument. And similarly for `RightCongruent`. Is that a bug? Or some special meaning of "left" and "right" that I'm missing?
https://github.com/agda/agda-stdlib/blob/650e05f3a5ef06cb938c3b91a4fc8b9b8d7c5ef2/src/Algebra/Definitions.agda#L33-L37

Also, I'm not a huge fan of left/right because it doesn't generalize to *n*-ary operations. How about `mono₁` and `mono₂`?

Reply:
Hmm, yeah, I'd have thought `congˡ` had type `∀ {x x′ y} → x ≈ x′ → x ∙ y ≈ x′ ∙ y` too. I *can* justify the current naming as saying that *left*-multiplication (by `x`) preserves equivalence, but when I'm using it in a proof, I don't think I'd think that way. IIRC, there's a similar problem with distributivity or zero, where one is the reverse of the other (something like `LeftZero` means `0 * x ≈ x` but `LeftDistributive` means `x * (y + z) ≈ x * y + x * z`, or vice versa).

As for numbers, I guess it side-steps any existing naming conventions, but introduces the problem of needing to choose between being 0-based or 1-based. I think (3+n)-ary operations are very rare in algebra, so I'm not too worried about that potential generalisation.

Reply:
OK, I can see that someone might have had good reasons to name `{Left,Right}Congruent` the way they did and I don't want to break backwards compatibility, so I'll not "fix" this unless @MatthewDaggitt tells me to. As for the `mono` variants, I'll go with `mono₁` and `mono₂` instead of for `monoˡ` and `monoʳ` for now, mostly because I don't know whether to follow that strange backwards terminology or not. (Again, unless @MatthewDaggitt tells me otherwise.)